Moved the disc model description to main text (#39)
greenelab/text_mined_hetnet_manuscript: content/03.methods.md

In the framework we used predictions from the generative model, $\hat{Y} = P(Y \mid \Lambda)$, as training classes for our dataset [@doi:10.14778/3157794.3157797; @doi:10.1145/3209889.3209898]. 

### Discriminative Model

> The discriminative model is a neural network, which we trained using predict labels from the generative model.

Reviewer: ```suggestion
The discriminative model is a neural network, which we trained using labels predicted from the generative model.
```

> The discriminative model is a neural network, which we trained using labels predicted from the generative model.

Reviewer 2: ```suggestion
The generative model produces predicted probabilities for each sentence by integrating labels.
The discriminative model is a neural network trained to produce the labels predicted from the generative model from sentence representations.
```

> The expectation is that the discriminative model can improve upon the generative model's predictions by incorporating textual features along with generated training labels.

Reviewer 2: ```suggestion
The goal of this combined approach is to develop models that learn text features associated with the overall task that go beyond the supplied label functions.
```

We used a piecewise convolutional neural network that contains multiple kernel filters as our discriminative model.

> This network uses multiple filters with fixed widths of 300 dimensions and a fixed height of 7 (Figure {@fig:convolutional_network}).

Reviewer 2: ```suggestion
We built a network with multiple filters using a fixed width of 300 dimensions and a fixed height of 7 (Figure {@fig:convolutional_network}).
```
Is dimensions the right term here - is this the window size for words?

Author: This is the window size for the words. Will change dimensions to something a bit more straight forward.

> We choose a fixed height of 7 because this height provided the best performance in terms of relationship classification [@arxiv:1510.03820].

Reviewer: It's not clear whether you or `@arxiv:1510.03820` found the fixed height of 7 was best?

> We choose a fixed height of 7 because this height was previously established to provide the best performance in relationship classification [@arxiv:1510.03820].

Reviewer 2: ```suggestion
We choose a fixed height of 7 because this height was previously reported to optimize performance in relationship classification [@arxiv:1510.03820].
```

> We trained this model for 15 epochs using the Adam optimizer [@arxiv:1412.6980] with pytorch's default parameter settings and a learning rate of 0.001 that decreases by half every epoch until the lower bound of 1e-5 is reached.

Reviewer 2: ```suggestion
We trained this model for 15 epochs using the Adam optimizer [@arxiv:1412.6980] with pytorch's default parameter settings and a learning rate of 0.001 that decreases by half every epoch until the lower bound of 1e-5 is reached, which we observed was often sufficient for convergence.
```
I think you showed me results that suggested that 15 epochs was sufficient by looking at convergence. Check to be sure. This is a bit of results-like text, but I think the point is unrelated enough to your main paper that it could go in here & it helps the reader understand why you stopped at 15.

Author: Sending a note to confirm that 15 epochs is sufficient for convergence.

The extracted features were concatenated into a singular vector that was passed into a fully connected network. 
The fully connected network had 300 neurons for the first layer, 100 neurons for the second layer and 50 neurons for the last layer. 
The last step from the fully connected network was to generate predictions using a softmax layer.
](images/figures/convolutional_neural_network/convolutional_neural_nework.png){#fig:convolutional_network}

#### Word Embeddings

Word embeddings are representations that map individual words to real valued vectors of user-specified dimensions.
These embeddings have been shown to capture the semantic and syntactic information between words [@arxiv:1310.4546].
We trained Facebook's fastText [@arxiv:1607.04606] using all candidate sentences for each individual relationship pair to generate word embeddings.
FastText uses a skip-gram model [@arxiv:1301.3781] that aims to predict the surrounding context for a candidate word and pairs the model with a novel scoring function that treats each word as a bag of character n-grams.
We trained this model for 20 epochs using a window size of 2 and generated 300-dimensional word embeddings.

> We use the optimized word embeddings to train a discriminative model. 

Reviewer 2: ```suggestion
We use the optimized word embeddings as input to our discriminative model. 
```

#### Calibration of the Discriminative Model

Often many tasks require a machine learning model to output reliable probability predictions. 
A model is well calibrated if the probabilities emitted from the model match the observed probabilities.
For example, a well-calibrated model that assigns a class label with 80% probability should have that class appear 80% of the time.
Deep neural network models can often be poorly calibrated [@arxiv:1706.04599; @arxiv:1807.00263].
These models are usually over-confident in their predictions.
For this reason, we calibrated our convolutional neural network using temperature scaling. 
Temperature scaling uses a parameter T to scale each value of the logit vector (z) before being passed into the softmax (SM) function.

$$\sigma_{SM}(\frac{z_{i}}{T}) = \frac{\exp(\frac{z_{i}}{T})}{\sum_{i}\exp(\frac{z_{i}}{T})}$$

We found the optimal T by minimizing the negative log likelihood (NLL) of the tune set.

> The benefit of using this method is that the model becomes more reliable and the accuracy of the model doesn't change [@arxiv:1706.04599].

Reviewer 2: ```suggestion
```
